rsyslog 8.2604.0 compile issue

### Discussed in https://github.com/rsyslog/rsyslog/discussions/6879

<div type='discussions-op-text'>

<sup>Originally posted by **LFrank2021** May 12, 2026</sup>
Hello,

I am trying to compile the above mentioned release for my AIX (IBM POWER) and am running into an issue:

My configure options:
```
export CC="xlclang++"
export PKG_CONFIG_PATH=/usr/lib/pkgconfig/
export CFLAGS="$CFLAGS -DSYSLOGD_PIDNAME=\\\"rsyslogd.pid\\\""
./configure \
    --prefix=%{_prefix} \
    --mandir=%{_mandir} \
    --infodir=%{_infodir} \
    --disable-uuid \
    --disable-libgcrypt \
    --disable-testbench \
    --disable-helgrind --disable-default-tests \
    --disable-fmhttp --disable-fmhash \
    --disable-libsystemd \
    --disable-impstats \
    --disable-impstats-push \
    --enable-openssl \
    --enable-omfile-hardened \
    --enable-omstdout \
    --enable-imfile \
    --with-systemdsystemunitdir=no \
    --enable-debug
```

resulting in the following make warnings/errors:
```
Making all in compat
  CC       compat_la-getifaddrs.lo
  CC       compat_la-strndup.lo
  CC       compat_la-asprintf.lo
  CC       compat_la-solaris_elf_fix.lo
asprintf.c:27:5: warning: no previous prototype for function 'asprintf' [-Wmissing-prototypes]
int asprintf(char **strp, const char *fmt, ...) {
    ^
1 warning generated.
getifaddrs.c:155:9: warning: implicit declaration of function 'close' is invalid in C99 [-Wimplicit-function-declaration]
        close(sock4);
        ^
getifaddrs.c:155:9: warning: this function declaration is not a prototype [-Wstrict-prototypes]
getifaddrs.c:174:16: warning: cast from 'char *' to 'struct ifreq *' increases required alignment from 1 to 8 [-Wcast-align]
        ifrp = (struct ifreq *)s_ifrp;
               ^~~~~~~~~~~~~~~~~~~~~~
getifaddrs.c:190:13: warning: implicit declaration of function 'ioctl' is invalid in C99 [-Wimplicit-function-declaration]
        if (ioctl(s, SIOCGIFFLAGS, (caddr_t)&ifrl) < 0) goto fail;
            ^
getifaddrs.c:190:13: warning: this function declaration is not a prototype [-Wstrict-prototypes]
getifaddrs.c:142:12: warning: unused variable 'n' [-Wunused-variable]
    int s, n, iflen;
           ^
getifaddrs.c:150:9: warning: unused variable 'flag' [-Wunused-variable]
    int flag;
        ^
getifaddrs.c:278:34: warning: unused parameter 'af' [-Wunused-parameter]
int getallifs(int s, sa_family_t af, struct ifreq **ifr, int *iflen, int64_t ifc_flags) {
                                 ^
getifaddrs.c:278:78: warning: unused parameter 'ifc_flags' [-Wunused-parameter]
int getallifs(int s, sa_family_t af, struct ifreq **ifr, int *iflen, int64_t ifc_flags) {
                                                                             ^
getifaddrs.c:305:16: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare]
    if (*iflen >= bufsize) {
        ~~~~~~ ^  ~~~~~~~
10 warnings generated.
  CCLD     compat.la
Target "all" is up to date.
Making all in grammar
        make  all-am
  CC       libgrammar_la-rainerscript.lo
  CC       libgrammar_la-grammar.lo
  CC       libgrammar_la-lexer.lo
In file included from rainerscript.c:43:
../runtime/rsyslog.h:49:40: warning: unknown warning group '-Wformat-truncation', ignored [-Wunknown-pragmas]
        #pragma GCC diagnostic ignored "-Wformat-truncation"
                                       ^
../runtime/rsyslog.h:59:40: warning: unknown warning group '-Wjump-misses-init', ignored [-Wunknown-pragmas]
        #pragma GCC diagnostic ignored "-Wjump-misses-init"
                                       ^
../runtime/rsyslog.h:61:40: warning: unknown warning group '-Wcast-function-type', ignored [-Wunknown-pragmas]
        #pragma GCC diagnostic ignored "-Wcast-function-type"
                                       ^
In file included from lexer.l:38:
../runtime/rsyslog.h:49:40: warning: unknown warning group '-Wformat-truncation', ignored [-Wunknown-pragmas]
        #pragma GCC diagnostic ignored "-Wformat-truncation"
                                       ^
../runtime/rsyslog.h:59:40: warning: unknown warning group '-Wjump-misses-init', ignored [-Wunknown-pragmas]
        #pragma GCC diagnostic ignored "-Wjump-misses-init"
                                       ^
../runtime/rsyslog.h:61:40: warning: unknown warning group '-Wcast-function-type', ignored [-Wunknown-pragmas]
        #pragma GCC diagnostic ignored "-Wcast-function-type"
                                       ^
In file included from rainerscript.c:48:
In file included from ../runtime/rsconf.h:29:
../runtime/dynstats.h:89:5: error: type name requires a specifier or qualifier
    STAILQ_HEAD(head, file_write_entry_s) q;
    ^
../runtime/dynstats.h:89:17: error: a parameter list without types is only allowed in a function definition
    STAILQ_HEAD(head, file_write_entry_s) q;
                ^
../runtime/dynstats.h:89:42: error: expected ';' at end of declaration list
    STAILQ_HEAD(head, file_write_entry_s) q;
                                         ^
                                         ;
3 warnings generated.
3 warnings and 3 errors generated.
Error while processing rainerscript.c.
```

Compiler is xlclang++ (16.1) last worked successfully with release 2508.

Any ideas and hints on how to solve this?

Best regards
Frank</div>

--- a/configure.ac
+++ b/configure.ac
@@ -211,7 +211,30 @@ AC_CHECK_HEADERS([malloc.h],[],[],[
      #endif
   ]
 ])
-AC_CHECK_HEADERS([fcntl.h locale.h netdb.h netinet/in.h paths.h stddef.h stdlib.h string.h sys/file.h sys/ioctl.h sys/param.h sys/socket.h sys/time.h sys/stat.h unistd.h utmp.h utmpx.h sys/epoll.h sys/prctl.h sys/select.h getopt.h linux/close_range.h linux/fs.h])
+AC_CHECK_HEADERS([fcntl.h locale.h netdb.h netinet/in.h paths.h stddef.h stdlib.h string.h sys/file.h sys/ioctl.h sys/param.h sys/socket.h sys/time.h sys/stat.h sys/queue.h unistd.h utmp.h utmpx.h sys/epoll.h sys/prctl.h sys/select.h getopt.h linux/close_range.h linux/fs.h])
+
+AC_MSG_CHECKING([for STAILQ macros in sys/queue.h])
+AC_COMPILE_IFELSE(
+    [AC_LANG_PROGRAM(
+        [[
+#include <sys/queue.h>
+struct stailq_test_entry {
+    STAILQ_ENTRY(stailq_test_entry) link;
+};
+STAILQ_HEAD(stailq_test_head, stailq_test_entry);
+        ]],
+        [[
+struct stailq_test_head head;
+STAILQ_INIT(&head);
+return STAILQ_EMPTY(&head) ? 0 : 1;
+        ]]
+    )],
+    [
+        AC_MSG_RESULT([yes])
+        AC_DEFINE([HAVE_SYS_QUEUE_STAILQ], [1], [Define if sys/queue.h provides STAILQ macros])
+    ],
+    [AC_MSG_RESULT([no])]
+)
 
 # Checks for typedefs, structures, and compiler characteristics.
 AC_C_CONST

--- a/plugins/imptcp/imptcp.c
+++ b/plugins/imptcp/imptcp.c
@@ -50,7 +50,7 @@
 #include <sys/socket.h>
 #include <sys/un.h>
 #include <sys/epoll.h>
-#include <sys/queue.h>
+#include "compat_queue.h"
 #include <netinet/tcp.h>
 #include <stdint.h>
 #include <zlib.h>

--- a/plugins/omazureeventhubs/omazureeventhubs.c
+++ b/plugins/omazureeventhubs/omazureeventhubs.c
@@ -30,7 +30,7 @@
 #include <fcntl.h>
 #include <pthread.h>
 #include <sys/uio.h>
-#include <sys/queue.h>
+#include "compat_queue.h"
 #include <sys/types.h>
 #include <math.h>
 #ifdef HAVE_SYS_STAT_H

--- a/plugins/omdtls/omdtls.c
+++ b/plugins/omdtls/omdtls.c
@@ -32,7 +32,7 @@
 #include <fcntl.h>
 #include <pthread.h>
 #include <sys/uio.h>
-#include <sys/queue.h>
+#include "compat_queue.h"
 #include <netdb.h>
 #include <sys/types.h>
 #include <sys/socket.h>

--- a/plugins/omkafka/omkafka.c
+++ b/plugins/omkafka/omkafka.c
@@ -31,7 +31,7 @@
 #include <fcntl.h>
 #include <pthread.h>
 #include <sys/uio.h>
-#include <sys/queue.h>
+#include "compat_queue.h"
 #include <sys/types.h>
 #include <math.h>
 #ifdef HAVE_SYS_STAT_H

--- a/runtime/Makefile.am
+++ b/runtime/Makefile.am
@@ -58,6 +58,7 @@ librsyslog_la_SOURCES = \
 	modules.h \
 	statsobj.c \
 	statsobj.h \
+	compat_queue.h \
 	dynstats.c \
 	dynstats.h \
 	perctile_ringbuf.c \

--- a/runtime/compat_queue.h
+++ b/runtime/compat_queue.h
@@ -1,0 +1,67 @@
+/*
+ * Compatibility helpers for BSD sys/queue.h users.
+ *
+ * Use the platform STAILQ macros when configure found them.  Some platforms
+ * provide sys/queue.h without that subset, so keep the local fallback limited
+ * to the macros used by rsyslog.
+ */
+#ifndef INCLUDED_COMPAT_QUEUE_H
+#define INCLUDED_COMPAT_QUEUE_H
+
+#include <stddef.h>
+#ifdef HAVE_SYS_QUEUE_H
+    #include <sys/queue.h>
+#endif
+
+#ifndef HAVE_SYS_QUEUE_STAILQ
+    #ifndef STAILQ_HEAD
+        #define STAILQ_HEAD(name, type)  \
+            struct name {                \
+                struct type *stqh_first; \
+                struct type **stqh_last; \
+            }
+    #endif
+
+    #ifndef STAILQ_ENTRY
+        #define STAILQ_ENTRY(type)      \
+            struct {                    \
+                struct type *stqe_next; \
+            }
+    #endif
+
+    #ifndef STAILQ_INIT
+        #define STAILQ_INIT(head)                        \
+            do {                                         \
+                (head)->stqh_first = NULL;               \
+                (head)->stqh_last = &(head)->stqh_first; \
+            } while (0)
+    #endif
+
+    #ifndef STAILQ_EMPTY
+        #define STAILQ_EMPTY(head) ((head)->stqh_first == NULL)
+    #endif
+
+    #ifndef STAILQ_FIRST
+        #define STAILQ_FIRST(head) ((head)->stqh_first)
+    #endif
+
+    #ifndef STAILQ_INSERT_TAIL
+        #define STAILQ_INSERT_TAIL(head, elm, field)         \
+            do {                                             \
+                (elm)->field.stqe_next = NULL;               \
+                *(head)->stqh_last = (elm);                  \
+                (head)->stqh_last = &(elm)->field.stqe_next; \
+            } while (0)
+    #endif
+
+    #ifndef STAILQ_REMOVE_HEAD
+        #define STAILQ_REMOVE_HEAD(head, field)                                           \
+            do {                                                                          \
+                if (((head)->stqh_first = (head)->stqh_first->field.stqe_next) == NULL) { \
+                    (head)->stqh_last = &(head)->stqh_first;                              \
+                }                                                                         \
+            } while (0)
+    #endif
+#endif
+
+#endif /* #ifndef INCLUDED_COMPAT_QUEUE_H */

--- a/runtime/dynstats.c
+++ b/runtime/dynstats.c
@@ -33,7 +33,7 @@
 #include "datetime.h"
 #include "unicode-helper.h"
 #include "hashtable_itr.h"
-#include <sys/queue.h>
+#include "compat_queue.h"
 
 /* definitions for objects we access */
 DEFobjStaticHelpers;

--- a/runtime/dynstats.h
+++ b/runtime/dynstats.h
@@ -18,7 +18,7 @@
 #ifndef INCLUDED_DYNSTATS_H
 #define INCLUDED_DYNSTATS_H
 
-#include <sys/queue.h>
+#include "compat_queue.h"
 #include "hashtable.h"
 
 typedef struct hashtable htable;

